File: code/dart/2-Async-programming/lib/async_comparison.js
function fetchUserData(userId) {
  return new Promise((resolve, reject) => {
    if (userId > 0) {
      // Success case
      setTimeout(() => {
        // Promise succeeds with this value
        resolve("John Doe"); 
      }, 1000);
    } else {
      // Failure case 
      // Promise fails with this error 
      reject(new Error("Invalid user ID")); 
    }
  });
}

async function main() {
  try {
    const user = await fetchUserData(123); //   Gets "John Doe"
    console.log(user);
  } catch (error) {
    console.log(error.message); // Handles any   rejection
  }
}  
main()

File: code/dart/2-Async-programming/lib/async_comparison.dart
Future<String> fetchUserData(int userId) async {
  if (userId > 0) {
    // Success case
    await Future.delayed(Duration(seconds: 1));
    // Future completes with this value
    return "John Doe";
  } else {
    // Failure case
    // Future completes with an error
    throw Exception("Invalid user ID");
  }
}

void main() async {
  try {
    final user = await fetchUserData(123); //   Gets "John Doe"
    print(user);
  } catch (error) {
    print(error.toString()); // Handles any   exception
  }
}